File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/gigs/quickpost/PostGigV1ViewModel.kt
@file:Suppress("MagicNumber", "TooManyFunctions")

package app.pantopus.android.ui.screens.gigs.quickpost

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import app.pantopus.android.data.api.models.gigs.CreateGigBody
import app.pantopus.android.data.api.models.gigs.CreateGigLocation
import app.pantopus.android.data.api.net.NetworkResult
import app.pantopus.android.data.files.FilesRepository
import app.pantopus.android.data.gigs.GigsRepository
import app.pantopus.android.ui.screens.gigs.GigsCategory
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.time.LocalDateTime
import java.time.ZoneId
import java.util.UUID
import javax.inject.Inject

enum class PostGigV1PriceType(
    val label: String,
    val unitLabel: String?,
) {
    Flat("Flat", "flat"),
    Hourly("Hourly", "/ hr"),
    Free("Free", null),
}

enum class PostGigV1PhotoTone { Sofa, Stairs, Street, Neutral }

/** P0.2 — per-tile upload lifecycle for the photo grid. */
enum class PostGigV1PhotoStatus { Uploading, Failed, Uploaded }

data class PostGigV1Photo(
    val id: String,
    val tone: PostGigV1PhotoTone,
    /** P0.2 — upload state; sample/preview tiles default to uploaded. */
    val status: PostGigV1PhotoStatus = PostGigV1PhotoStatus.Uploaded,
    /** P0.2 — backend URL once the upload lands; rides `attachments`. */
    val url: String? = null,
)

data class PostGigV1Form(
    val category: GigsCategory = GigsCategory.All,
    val title: String = "",
    val description: String = "",
    val price: String = "",
    val priceType: PostGigV1PriceType = PostGigV1PriceType.Flat,
    val scheduledAt: LocalDateTime = LocalDateTime.now().plusDays(1),
    val location: String = "",
    val photos: List<PostGigV1Photo> = emptyList(),
) {
    val hasAnyInput: Boolean
        get() =
            category != GigsCategory.All ||
                title.isNotBlank() ||
                description.isNotBlank() ||
                price.isNotBlank() ||
                priceType != PostGigV1PriceType.Flat ||
                location.isNotBlank() ||
                photos.isNotEmpty()
}

enum class PostGigV1Field { Category, Title, Description, Price, DateTime, Location }

data class PostGigV1ValidationError(
    val field: PostGigV1Field,
    val message: String,
)

sealed interface PostGigV1UiState {
    data object Loading : PostGigV1UiState

    data object Empty : PostGigV1UiState

    data class Content(
        val form: PostGigV1Form = PostGigV1Form(),
        val validationErrors: List<PostGigV1ValidationError> = emptyList(),
        val isSubmitting: Boolean = false,
        val postedGigId: String? = null,
    ) : PostGigV1UiState {
        /** P0.2 — true while any photo upload is still in flight. */
        val hasUploadsInFlight: Boolean = form.photos.any { it.status == PostGigV1PhotoStatus.Uploading }
        val canAttemptSubmit: Boolean = !isSubmitting && !hasUploadsInFlight
        val isPostEnabled: Boolean = canAttemptSubmit && (form.hasAnyInput || validationErrors.isNotEmpty())
    }

    data class FatalError(
        val message: String,
    ) : PostGigV1UiState
}

sealed interface PostGigV1Event {
    data class Posted(
        val gigId: String,
    ) : PostGigV1Event
}

@HiltViewModel
class PostGigV1ViewModel
    @Inject
    constructor(
        private val repo: GigsRepository,
        private val filesRepo: FilesRepository,
    ) : ViewModel() {
        private val _state = MutableStateFlow<PostGigV1UiState>(PostGigV1UiState.Content())
        val state: StateFlow<PostGigV1UiState> = _state.asStateFlow()

        private val _pendingEvent = MutableStateFlow<PostGigV1Event?>(null)
        val pendingEvent: StateFlow<PostGigV1Event?> = _pendingEvent.asStateFlow()

        // Stashed so a post failure can flip to FatalError yet restore the
        // filled form on retry (mirrors iOS, where the form survives .error).
        private var lastForm: PostGigV1Form? = null

        // P0.2 — picked-photo bytes (for retry) + per-tile upload jobs.
        private val pendingPhotoBytes = mutableMapOf<String, PostGigV1PickedPhoto>()
        private val uploadJobs = mutableMapOf<String, Job>()

        fun acknowledgeEvent() {
            _pendingEvent.value = null
        }

        fun retry() {
            _state.value = PostGigV1UiState.Content(form = lastForm ?: PostGigV1Form())
            lastForm = null
        }

        fun startFromEmpty() {
            _state.value = PostGigV1UiState.Content()
        }

        fun preselectCategoryIfNeeded(category: GigsCategory) {
            updateContent { content ->
                if (content.form.category != GigsCategory.All || category == GigsCategory.All) {
                    content
                } else {
                    content.copy(form = content.form.copy(category = category))
                }
            }
        }

        fun updateCategory(category: GigsCategory) {
            updateForm { it.copy(category = category) }
        }

        fun updateTitle(title: String) {
            updateForm { it.copy(title = title) }
        }

        fun updateDescription(description: String) {
            updateForm {
                it.copy(description = description.take(PostGigV1SampleData.DESCRIPTION_MAX_LENGTH))
            }
        }

        fun updatePrice(price: String) {
            updateForm {
                it.copy(price = price.filter { char -> char.isDigit() || char == '.' })
            }
        }

        fun updatePriceType(type: PostGigV1PriceType) {
            updateForm {
                it.copy(priceType = type, price = if (type == PostGigV1PriceType.Free) "" else it.price)
            }
        }

        fun updateScheduledAt(date: LocalDateTime) {
            updateForm { it.copy(scheduledAt = date) }
        }

        fun updateLocation(location: String) {
            updateForm { it.copy(location = location) }
        }

        /**
         * P0.2 — accept a picked photo and upload it immediately via
         * `POST /api/files/upload` (`FilesRepository`). The tile tracks
         * uploading / failed (tap-to-retry) / uploaded-URL states.
         */
        fun addPickedPhoto(picked: PostGigV1PickedPhoto) {
            val content = _state.value as? PostGigV1UiState.Content ?: return
            if (content.form.photos.size >= PostGigV1SampleData.MAX_PHOTOS) return
            val id = "photo-${UUID.randomUUID()}"
            pendingPhotoBytes[id] = picked
            updateForm { form ->
                form.copy(
                    photos =
                        form.photos +
                            PostGigV1Photo(
                                id = id,
                                tone = PostGigV1PhotoTone.Neutral,
                                status = PostGigV1PhotoStatus.Uploading,
                            ),
                )
            }
            startUpload(id)
        }

        /** P0.2 — retry a failed upload tile (bytes are still held). */
        fun retryPhotoUpload(id: String) {
            if (pendingPhotoBytes[id] == null) return
            updatePhoto(id) { it.copy(status = PostGigV1PhotoStatus.Uploading) }
            startUpload(id)
        }

        fun removePhoto(id: String) {
            uploadJobs.remove(id)?.cancel()
            pendingPhotoBytes.remove(id)
            updateForm { it.copy(photos = it.photos.filterNot { photo -> photo.id == id }) }
        }

        private fun startUpload(id: String) {
            val picked = pendingPhotoBytes[id] ?: return
            uploadJobs[id] =
                viewModelScope.launch {
                    val result =
                        filesRepo.uploadFile(
                            filename = picked.filename,
                            mimeType = picked.mimeType,
                            bytes = picked.bytes,
                            fileType = GIG_PHOTO_FILE_TYPE,
                            visibility = "public",
                        )
                    when (result) {
                        is NetworkResult.Success -> {
                            pendingPhotoBytes.remove(id)
                            uploadJobs.remove(id)
                            updatePhoto(id) {
                                it.copy(status = PostGigV1PhotoStatus.Uploaded, url = result.data.file.url)
                            }
                        }
                        is NetworkResult.Failure ->
                            updatePhoto(id) { it.copy(status = PostGigV1PhotoStatus.Failed) }
                    }
                }
        }

        private fun updatePhoto(
            id: String,
            transform: (PostGigV1Photo) -> PostGigV1Photo,
        ) {
            updateForm { form ->
                form.copy(photos = form.photos.map { if (it.id == id) transform(it) else it })
            }
        }

        /**
         * Validate, then create the gig via `POST /api/gigs`
         * (`GigsRepository.create`) — the same create path the V2 composer
         * and gigs feed use. On success the backend-issued id drives the
         * Posted event; on failure we flip to FatalError (the form is stashed
         * so retry restores it).
         */
        fun submit(now: LocalDateTime = LocalDateTime.now()) {
            val current = _state.value as? PostGigV1UiState.Content ?: return
            // P0.2 — never race a half-done upload; the Post CTA is also
            // disabled while uploads are in flight.
            if (!current.canAttemptSubmit) return
            val errors = validate(current.form, now)
            if (errors.isNotEmpty()) {
                _state.value = current.copy(validationErrors = errors)
                return
            }
            _state.value = current.copy(validationErrors = emptyList(), isSubmitting = true)
            viewModelScope.launch {
                when (val result = repo.create(buildCreateBody(current.form))) {
                    is NetworkResult.Success -> {
                        val gigId = result.data.gig.id
                        _state.value = current.copy(isSubmitting = false, postedGigId = gigId)
                        _pendingEvent.value = PostGigV1Event.Posted(gigId)
                    }
                    is NetworkResult.Failure -> {
                        lastForm = current.form
                        _state.value = PostGigV1UiState.FatalError(result.error.message)
                    }
                }
            }
        }

        /**
         * Map the V1 form onto the `POST /api/gigs` body. The legacy composer
         * collects a free-text location only, so it rides as the `custom`
         * location `address` with a `(0, 0)` placeholder coordinate. Pay-type
         * maps Flat→`fixed`, Hourly→`hourly`, Free→`offers`; the backend
         * rejects a non-positive price so Free uses a `1` sentinel.
         */
        private fun buildCreateBody(form: PostGigV1Form): CreateGigBody {
            val trimmedPrice = form.price.trim().toDoubleOrNull() ?: 0.0
            val payType: String
            val price: Double
            when (form.priceType) {
                PostGigV1PriceType.Flat -> {
                    payType = "fixed"
                    price = if (trimmedPrice > 0.0) trimmedPrice else 1.0
                }
                PostGigV1PriceType.Hourly -> {
                    payType = "hourly"
                    price = if (trimmedPrice > 0.0) trimmedPrice else 1.0
                }
                PostGigV1PriceType.Free -> {
                    payType = "offers"
                    price = 1.0
                }
            }
            return CreateGigBody(
                title = form.title.trim(),
                description = form.description.trim(),
                category = if (form.category == GigsCategory.All) null else form.category.key,
                price = price,
                payType = payType,
                scheduleType = "scheduled",
                scheduledStart = form.scheduledAt.atZone(ZoneId.systemDefault()).toInstant().toString(),
                taskFormat = null,
                // P0.2 — uploaded photo URLs ride as attachments.
                attachments = form.photos.mapNotNull { it.url }.ifEmpty { null },
                location =
                    CreateGigLocation(
                        mode = "custom",
                        latitude = 0.0,
                        longitude = 0.0,
                        address = form.location.trim(),
                    ),
            )
        }

        private fun updateForm(transform: (PostGigV1Form) -> PostGigV1Form) {
            updateContent { content -> content.copy(form = transform(content.form)) }
        }

        private fun updateContent(transform: (PostGigV1UiState.Content) -> PostGigV1UiState.Content) {
            val current = _state.value as? PostGigV1UiState.Content ?: return
            _state.value = transform(current)
        }

        private fun validate(
            form: PostGigV1Form,
            now: LocalDateTime,
        ): List<PostGigV1ValidationError> {
            val errors = mutableListOf<PostGigV1ValidationError>()
            if (form.category == GigsCategory.All) {
                errors += PostGigV1ValidationError(PostGigV1Field.Category, "Choose a category.")
            }
            if (form.title.isBlank()) {
                errors += PostGigV1ValidationError(PostGigV1Field.Title, "Title is required.")
            }
            if (form.description.trim().length < PostGigV1SampleData.DESCRIPTION_MIN_LENGTH) {
                errors +=
                    PostGigV1ValidationError(
                        PostGigV1Field.Description,
                        "Description must be at least ${PostGigV1SampleData.DESCRIPTION_MIN_LENGTH} characters.",
                    )
            }
            if (form.priceType != PostGigV1PriceType.Free) {
                val price = form.price.trim()
                if (price.isEmpty()) {
                    errors += PostGigV1ValidationError(PostGigV1Field.Price, "Enter a price, or pick Free.")
                } else if ((price.toDoubleOrNull() ?: 0.0) <= 0.0) {
                    errors += PostGigV1ValidationError(PostGigV1Field.Price, "Price must be greater than zero.")
                }
            }
            if (!form.scheduledAt.isAfter(now)) {
                errors += PostGigV1ValidationError(PostGigV1Field.DateTime, "Date is in the past. Pick a future time.")
            }
            if (form.location.isBlank()) {
                errors += PostGigV1ValidationError(PostGigV1Field.Location, "Add a pickup or meetup location.")
            }
            return errors
        }

        private companion object {
            /** P0.2 — `file_type` form field on `POST /api/files/upload`. */
            const val GIG_PHOTO_FILE_TYPE = "gig_photo"
        }
    }

/**
 * P0.2 — raw bytes of a picked photo, held by the view-model for upload +
 * retry. Not a data class — [bytes] is an array, so structural equality
 * would be misleading.
 */
class PostGigV1PickedPhoto(
    val filename: String,
    val mimeType: String,
    val bytes: ByteArray,
)

File: frontend/apps/ios/Pantopus/Features/Gigs/QuickPost/PostGigV1ViewModel.swift
//
//  PostGigV1ViewModel.swift
//  Pantopus
//
//  A13.8 — legacy single-screen gig composer. `submit()` posts the form to
//  `POST /api/gigs` (`GigsEndpoints.create`) — the same create path the V2
//  "Magic" composer (`GigComposeViewModel`) and the gigs feed use. The V2
//  `/api/gigs/magic-*` draft flow is separate and untouched here.
//
//  Phase 4 (A13.8 polish): photos ride the real `POST /api/files/upload`
//  pipeline (per-tile uploading / failed-retry / uploaded states, same
//  mechanism as the V2 wizard's P15.5 handling), and an optional
//  `editGigId` flips the screen into edit mode — `GET /api/gigs/:id`
//  prefills every field and submit goes to `PATCH /api/gigs/:id`.
//

import Foundation
import Observation

// swiftlint:disable file_length

public enum PostGigV1PriceType: String, CaseIterable, Identifiable, Sendable {
    case flat
    case hourly
    case free

    public var id: String {
        rawValue
    }

    public var label: String {
        switch self {
        case .flat: "Flat"
        case .hourly: "Hourly"
        case .free: "Free"
        }
    }

    public var unitLabel: String? {
        switch self {
        case .flat: "flat"
        case .hourly: "/ hr"
        case .free: nil
        }
    }
}

/// One photo tile riding the real upload pipeline
/// (`POST /api/files/upload`, field `file`, `file_type: "gig_photo"`).
/// The raw bytes back the grid thumbnail; `status` drives the per-tile
/// spinner / retry / uploaded chrome. Edit-mode prefill rehydrates tiles
/// from the gig's stored `attachments` (URL only, no bytes).
public struct PostGigV1Photo: Identifiable, Equatable, Sendable {
    public enum Status: Equatable, Sendable {
        case uploading
        case failed
        case uploaded(url: String)
    }

    public let id: String
    public let imageData: Data
    public var status: Status

    public init(id: String, imageData: Data = Data(), status: Status) {
        self.id = id
        self.imageData = imageData
        self.status = status
    }

    public var uploadedURL: String? {
        if case let .uploaded(url) = status { return url }
        return nil
    }
}

public struct PostGigV1Form: Equatable, Sendable {
    public var category: GigsCategory
    public var title: String
    public var description: String
    public var price: String
    public var priceType: PostGigV1PriceType
    public var scheduledAt: Date
    public var location: String
    public var photos: [PostGigV1Photo]

    public init(
        category: GigsCategory = .all,
        title: String = "",
        description: String = "",
        price: String = "",
        priceType: PostGigV1PriceType = .flat,
        scheduledAt: Date = Date().addingTimeInterval(86400),
        location: String = "",
        photos: [PostGigV1Photo] = []
    ) {
        self.category = category
        self.title = title
        self.description = description
        self.price = price
        self.priceType = priceType
        self.scheduledAt = scheduledAt
        self.location = location
        self.photos = photos
    }

    public var hasAnyInput: Bool {
        category != .all ||
            !title.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ||
            !description.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ||
            !price.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ||
            priceType != .flat ||
            !location.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ||
            !photos.isEmpty
    }
}

public enum PostGigV1Field: String, Sendable {
    case category
    case title
    case description
    case price
    case dateTime
    case location
}

public struct PostGigV1ValidationError: Identifiable, Equatable, Sendable {
    public let field: PostGigV1Field
    public let message: String

    public var id: String {
        field.rawValue
    }
}

public enum PostGigV1LoadState: Equatable, Sendable {
    case loading
    case empty
    case ready
    case error(String)
}

public struct PostGigV1State: Equatable, Sendable {
    public var loadState: PostGigV1LoadState
    public var form: PostGigV1Form
    public var validationErrors: [PostGigV1ValidationError]
    public var isSubmitting: Bool
    public var postedGigId: String?

    public init(
        loadState: PostGigV1LoadState = .ready,
        form: PostGigV1Form = PostGigV1Form(),
        validationErrors: [PostGigV1ValidationError] = [],
        isSubmitting: Bool = false,
        postedGigId: String? = nil
    ) {
        self.loadState = loadState
        self.form = form
        self.validationErrors = validationErrors
        self.isSubmitting = isSubmitting
        self.postedGigId = postedGigId
    }
}

@Observable
@MainActor
public final class PostGigV1ViewModel {
    public private(set) var state: PostGigV1State

    private let api: APIClient
    private let uploader: MultipartUploader
    private let referenceNow: Date?

    /// When set the screen is the "Edit gig" surface: `load()` prefills
    /// from `GET /api/gigs/:id` and `submit()` goes to `PATCH`.
    public let editGigId: String?

    /// One-shot guard around the edit-mode prefill fetch.
    private var editLoaded = false

    /// Exact coordinates captured at edit-load. PATCHing `location` with
    /// the V1 free-text address rides these so the stored point survives
    /// (the backend requires lat/lng on the nested location object).
    private var editOrigin: (latitude: Double, longitude: Double)?

    /// In-flight photo uploads keyed by photo id.
    private var uploadTasks: [String: Task<Void, Never>] = [:]

    /// Public entry point — carries no `APIClient` (the client and `.shared`
    /// are module-internal) so views / previews / sample data construct the
    /// composer without referencing it.
    public convenience init(
        initialState: PostGigV1State = PostGigV1State(),
        referenceNow: Date? = nil,
        editGigId: String? = nil
    ) {
        self.init(
            api: .shared,
            initialState: initialState,
            referenceNow: referenceNow,
            editGigId: editGigId
        )
    }

    /// Designated init — module-internal because `APIClient` is. Tests
    /// inject a stubbed client + uploader here.
    init(
        api: APIClient,
        uploader: MultipartUploader = .shared,
        initialState: PostGigV1State = PostGigV1State(),
        referenceNow: Date? = nil,
        editGigId: String? = nil
    ) {
        self.api = api
        self.uploader = uploader
        state = initialState
        self.referenceNow = referenceNow
        self.editGigId = editGigId
        if editGigId != nil {
            state.loadState = .loading
        }
    }

    public var isEditMode: Bool {
        editGigId != nil
    }

    /// Top-bar title — "Post gig" for create, "Edit gig" for edit.
    public var screenTitle: String {
        isEditMode ? "Edit gig" : "Post gig"
    }

    /// Right-action CTA label — "Post" for create, "Save" for edit.
    public var commitLabel: String {
        isEditMode ? "Save" : "Post"
    }

    /// True while any photo upload is still in flight — gates the
    /// Post / Save CTA so a half-uploaded gig can't ship.
    public var hasUploadsInFlight: Bool {
        state.form.photos.contains { $0.status == .uploading }
    }

    public var isPostEnabled: Bool {
        canAttemptSubmit && (state.form.hasAnyInput || !state.validationErrors.isEmpty)
    }

    public var canAttemptSubmit: Bool {
        state.loadState == .ready && !state.isSubmitting && !hasUploadsInFlight
    }

    /// Edit mode: fetch the gig and prefill every field. Create mode:
    /// no-op (the form starts ready).
    public func load() async {
        guard let editGigId, !editLoaded else { return }
        state.loadState = .loading
        do {
            let response: GigDetailResponse = try await api.request(GigsEndpoints.detail(id: editGigId))
            prefill(from: response.gig)
            editLoaded = true
            state.loadState = .ready
        } catch {
            let message = (error as? APIError)?.errorDescription ?? "Couldn't load this gig. Please try again."
            state.loadState = .error(message)
        }
    }

    public func retry() {
        if isEditMode, !editLoaded {
            // The edit prefill never landed — refetch instead of showing
            // an empty form against a PATCH submit.
            Task { await load() }
        } else {
            state.loadState = .ready
        }
    }

    public func startFromEmpty() {
        state = PostGigV1State(form: PostGigV1Form())
    }

    public func seedFilledSample() {
        state = PostGigV1State(form: PostGigV1SampleData.filledForm)
    }

    public func updateCategory(_ category: GigsCategory) {
        state.form.category = category
    }

    public func updateTitle(_ title: String) {
        state.form.title = title
    }

    public func updateDescription(_ description: String) {
        state.form.description = String(description.prefix(PostGigV1SampleData.descriptionMaxLength))
    }

    public func updatePrice(_ price: String) {
        let filtered = price.filter { $0.isNumber || $0 == "." }
        state.form.price = filtered
    }

    public func updatePriceType(_ priceType: PostGigV1PriceType) {
        state.form.priceType = priceType
        if priceType == .free {
            state.form.price = ""
        }
    }

    public func updateScheduledAt(_ date: Date) {
        state.form.scheduledAt = date
    }

    public func updateLocation(_ location: String) {
        state.form.location = location
    }

    // MARK: - Photo uploads (same pipeline as the V2 wizard's P15.5)

    /// Add a picked photo and immediately upload it in the background.
    /// Caps the grid at `PostGigV1SampleData.maxPhotos` — extra calls
    /// are ignored. The first photo is the gig's cover.
    public func addPhotoData(_ data: Data) {
        guard state.form.photos.count < PostGigV1SampleData.maxPhotos, !data.isEmpty else { return }
        let photo = PostGigV1Photo(id: UUID().uuidString, imageData: data, status: .uploading)
        state.form.photos.append(photo)
        startUpload(photoId: photo.id)
    }

    /// Tap-to-retry on a failed tile.
    public func retryUpload(id: String) {
        guard let index = state.form.photos.firstIndex(where: { $0.id == id }),
              state.form.photos[index].status == .failed else { return }
        state.form.photos[index].status = .uploading
        startUpload(photoId: id)
    }

    /// Remove a photo (any state). Cancels an in-flight upload.
    public func removePhoto(id: String) {
        uploadTasks[id]?.cancel()
        uploadTasks[id] = nil
        state.form.photos.removeAll { $0.id == id }
    }

    private func startUpload(photoId: String) {
        uploadTasks[photoId] = Task { [weak self] in
            await self?.performUpload(photoId: photoId)
        }
    }

    /// Push one photo through `POST /api/files/upload` (field `file`,
    /// `file_type: "gig_photo"` — the V2 wizard's exact mechanism). The
    /// resulting URL rides the create/update body's `attachments`.
    func performUpload(photoId: String) async {
        guard let photo = state.form.photos.first(where: { $0.id == photoId }) else { return }
        do {
            let response = try await uploader.uploadFile(
                MultipartFile(
                    fieldName: "file",
                    filename: "gig-\(photoId.prefix(6)).jpg",
                    mimeType: "image/jpeg",
                    data: photo.imageData
                ),
                formFields: ["file_type": "gig_photo"]
            )
            guard let index = state.form.photos.firstIndex(where: { $0.id == photoId }) else { return }
            state.form.photos[index].status = .uploaded(url: response.file.url)
        } catch {
            guard let index = state.form.photos.firstIndex(where: { $0.id == photoId }) else { return }
            state.form.photos[index].status = .failed
        }
    }

    #if DEBUG
    /// Test hook — wait for every kicked upload task to settle.
    func awaitUploadsForTesting() async {
        for task in uploadTasks.values {
            await task.value
        }
    }
    #endif

    // MARK: - Submit

    /// Validate, then create (`POST /api/gigs`) or — in edit mode —
    /// update (`PATCH /api/gigs/:id`) the gig. On success the
    /// backend-issued gig id is stored in `state.postedGigId` and
    /// returned so the caller can route to the task; on failure the
    /// message surfaces via `loadState = .error(_)` and `retry()`
    /// returns to the still-filled form.
    @discardableResult
    public func submit() async -> String? {
        guard canAttemptSubmit else { return nil }
        let errors = validate(form: state.form)
        guard errors.isEmpty else {
            state.validationErrors = errors
            return nil
        }
        state.validationErrors = []
        state.isSubmitting = true
        defer { state.isSubmitting = false }
        do {
            let response: CreateGigResponse = if let editGigId {
                try await api.request(
                    GigsEndpoints.update(id: editGigId, body: buildUpdateBody(from: state.form))
                )
            } else {
                try await api.request(
                    GigsEndpoints.create(buildCreateBody(from: state.form))
                )
            }
            state.postedGigId = response.gig.id
            return response.gig.id
        } catch {
            let fallback = isEditMode
                ? "Couldn't save your changes. Please try again."
                : "Couldn't post your task. Please try again."
            let message = (error as? APIError)?.errorDescription ?? fallback
            state.loadState = .error(message)
            return nil
        }
    }

    public func error(for field: PostGigV1Field) -> String? {
        state.validationErrors.first { $0.field == field }?.message
    }

    public func dateLabel(for date: Date) -> String {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.dateFormat = "EEE, MMM d · h:mm a"
        return formatter.string(from: date)
    }
}

// MARK: - Body building, validation, edit prefill

extension PostGigV1ViewModel {
    /// Pay-type maps flat→`fixed`, hourly→`hourly`, free→`offers` with a
    /// true `price: 0` — the backend schema accepts zero
    /// (`Joi.number().min(0)`, `backend/routes/gigs.js:428`).
    private func payTypeAndPrice(from form: PostGigV1Form) -> (payType: String, price: Double) {
        let trimmedPrice = Double(form.price.trimmingCharacters(in: .whitespacesAndNewlines)) ?? 0
        switch form.priceType {
        case .flat:
            return ("fixed", trimmedPrice > 0 ? trimmedPrice : 1)
        case .hourly:
            return ("hourly", trimmedPrice > 0 ? trimmedPrice : 1)
        case .free:
            return ("offers", 0)
        }
    }

    private func uploadedAttachmentURLs(from form: PostGigV1Form) -> [String] {
        form.photos.compactMap(\.uploadedURL)
    }

    /// Map the V1 form onto the `POST /api/gigs` body. The legacy composer
    /// collects a free-text location only, so we send it as the `custom`
    /// location `address` with a `(0, 0)` placeholder coordinate — the same
    /// fallback the V2 composer uses when it has no geocode
    /// (`GigComposeViewModel.fallbackLocation`).
    private func buildCreateBody(from form: PostGigV1Form) -> CreateGigBody {
        let pay = payTypeAndPrice(from: form)
        let attachments = uploadedAttachmentURLs(from: form)
        return CreateGigBody(
            title: form.title.trimmingCharacters(in: .whitespacesAndNewlines),
            description: form.description.trimmingCharacters(in: .whitespacesAndNewlines),
            category: form.category == .all ? nil : form.category.rawValue,
            price: pay.price,
            payType: pay.payType,
            scheduleType: "scheduled",
            scheduledStart: ISO8601DateFormatter().string(from: form.scheduledAt),
            taskFormat: nil,
            attachments: attachments.isEmpty ? nil : attachments,
            location: CreateGigLocation(
                mode: "custom",
                latitude: 0,
                longitude: 0,
                address: form.location.trimmingCharacters(in: .whitespacesAndNewlines)
            )
        )
    }

    /// Map the V1 form onto the `PATCH /api/gigs/:id` body — same field
    /// names as create. `attachments` always rides (an empty array
    /// clears removed photos); `location` only rides when edit-load
    /// captured real coordinates, so the stored point is preserved.
    private func buildUpdateBody(from form: PostGigV1Form) -> UpdateGigBody {
        let pay = payTypeAndPrice(from: form)
        var location: CreateGigLocation?
        if let editOrigin {
            location = CreateGigLocation(
                mode: "custom",
                latitude: editOrigin.latitude,
                longitude: editOrigin.longitude,
                address: form.location.trimmingCharacters(in: .whitespacesAndNewlines)
            )
        }
        return UpdateGigBody(
            title: form.title.trimmingCharacters(in: .whitespacesAndNewlines),
            description: form.description.trimmingCharacters(in: .whitespacesAndNewlines),
            category: form.category == .all ? nil : form.category.rawValue,
            price: pay.price,
            payType: pay.payType,
            scheduleType: "scheduled",
            scheduledStart: ISO8601DateFormatter().string(from: form.scheduledAt),
            attachments: uploadedAttachmentURLs(from: form),
            location: location
        )
    }

    private func validate(form: PostGigV1Form) -> [PostGigV1ValidationError] {
        var errors: [PostGigV1ValidationError] = []
        if form.category == .all {
            errors.append(.init(field: .category, message: "Choose a category."))
        }
        if form.title.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            errors.append(.init(field: .title, message: "Title is required."))
        }
        if form.description.trimmingCharacters(in: .whitespacesAndNewlines).count < PostGigV1SampleData.descriptionMinLength {
            errors.append(.init(
                field: .description,
                message: "Description must be at least \(PostGigV1SampleData.descriptionMinLength) characters."
            ))
        }
        if form.priceType != .free {
            let trimmed = form.price.trimmingCharacters(in: .whitespacesAndNewlines)
            if trimmed.isEmpty {
                errors.append(.init(field: .price, message: "Enter a price, or pick Free."))
            } else if (Double(trimmed) ?? 0) <= 0 {
                errors.append(.init(field: .price, message: "Price must be greater than zero."))
            }
        }
        if form.scheduledAt <= (referenceNow ?? Date()) {
            errors.append(.init(field: .dateTime, message: "Date is in the past. Pick a future time."))
        }
        if form.location.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            errors.append(.init(field: .location, message: "Add a pickup or meetup location."))
        }
        return errors
    }

    // MARK: - Edit-mode prefill

    private func prefill(from gig: GigDTO) {
        var form = PostGigV1Form()
        form.category = GigsCategory(rawValue: gig.category ?? "") ?? .all
        form.title = gig.title
        form.description = String((gig.description ?? "").prefix(PostGigV1SampleData.descriptionMaxLength))
        switch gig.payType {
        case "offers":
            form.priceType = .free
            form.price = ""
        case "hourly":
            form.priceType = .hourly
            form.price = Self.priceText(gig.price)
        default:
            form.priceType = .flat
            form.price = Self.priceText(gig.price)
        }
        if let scheduledStart = gig.scheduledStart, let date = Self.parseISO(scheduledStart) {
            form.scheduledAt = date
        }
        form.location = gig.exactAddress ?? gig.pickupAddress ?? ""
        form.photos = (gig.attachments ?? [])
            .prefix(PostGigV1SampleData.maxPhotos)
            .map { PostGigV1Photo(id: $0, status: .uploaded(url: $0)) }
        if let latitude = gig.location?.latitude ?? gig.latitude,
           let longitude = gig.location?.longitude ?? gig.longitude {
            editOrigin = (latitude, longitude)
        }
        state.form = form
        state.validationErrors = []
    }

    private static func priceText(_ price: Double?) -> String {
        guard let price, price > 0 else { return "" }
        if price.truncatingRemainder(dividingBy: 1) == 0 {
            return String(Int(price))
        }
        return String(price)
    }

    private static func parseISO(_ string: String) -> Date? {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let date = formatter.date(from: string) { return date }
        formatter.formatOptions = [.withInternetDateTime]
        return formatter.date(from: string)
    }
}
